File: src/2025-02-24-deepclaude-rust/python/exercises/practice/go-counting/go_counting.py
WHITE = 'W'
BLACK = 'B'
NONE = ''

class Board:
    """Count territories of each player in a Go game

    Args:
        board (list[str]): A two-dimensional Go board
    """

    def __init__(self, board):
        self.board = board

    def territory(self, x, y):
        """Find the owner and the territories given a coordinate on
           the board

        Args:
            x (int): Column on the board
            y (int): Row on the board

        Returns:
            (str, set): A tuple, the first element being the owner
                        of that area.  One of "W", "B", "".  The
                        second being a set of coordinates, representing
                        the owner's territories.
        """
        if not self.board:
            raise ValueError('Invalid coordinate')
        height = len(self.board)
        if height == 0:
            raise ValueError('Invalid coordinate')
        width = len(self.board[0])
        if x < 0 or x >= width or y < 0 or y >= height:
            raise ValueError('Invalid coordinate')
        
        # Check if current cell is a stone
        current_cell = self.board[y][x]
        if current_cell != ' ':
            return ('', set())
        
        from collections import deque
        visited = set()
        queue = deque([(x, y)])
        visited.add((x, y))
        territory = set()
        borders = set()
        touches_edge = False
        
        while queue:
            cx, cy = queue.popleft()
            territory.add((cx, cy))
            # Check all four directions
            for dx, dy in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
                nx = cx + dx
                ny = cy + dy
                if 0 <= nx < width and 0 <= ny < height:
                    neighbor_cell = self.board[ny][nx]
                    if neighbor_cell == ' ':
                        if (nx, ny) not in visited:
                            visited.add((nx, ny))
                            queue.append((nx, ny))
                    else:
                        if neighbor_cell in ('B', 'W'):
                            borders.add(neighbor_cell)
                else:
                    touches_edge = True
        
        owner = ''
        if not touches_edge:
            if len(borders) == 1:
                owner = borders.pop()
        return (owner, territory)

    def territories(self):
        """Find the owners and the territories of the whole board

        Args:
            none

        Returns:
            dict(str, set): A dictionary whose key being the owner
                        , i.e. "W", "B", "".  The value being a set
                        of coordinates owned by the owner.
        """
        territories = {'B': set(), 'W': set(), '': set()}
        visited = set()
        
        for y in range(len(self.board)):
            row = self.board[y]
            for x in range(len(row)):
                if (x, y) not in visited and row[x] == ' ':
                    owner, coords = self.territory(x, y)
                    territories[owner].update(coords)
                    visited.update(coords)
        
        return territories

File: src/2025-02-24-deepclaude-rust/javascript/exercises/practice/go-counting/go-counting.js
//
// This is only a SKELETON file for the 'Go Counting' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class GoCounting {
  constructor(board) {
    this.board = board.map(row => row.split(''));
    this.height = this.board.length;
    this.width = this.height > 0 ? this.board[0].length : 0;
  }

  getTerritory(x, y) {
    if (x < 0 || x >= this.width || y < 0 || y >= this.height) {
      return { owner: 'NONE', territory: [] };
    }

    const cell = this.board[y][x];
    if (cell !== ' ') {
      return { owner: 'NONE', territory: [] };
    }

    const visited = new Set();
    const queue = [[x, y]];
    visited.add(`${x},${y}`);
    let borders = new Set();
    let hasEdge = false;

    while (queue.length > 0) {
      const [currentX, currentY] = queue.shift();
      const directions = [[0, 1], [0, -1], [1, 0], [-1, 0]];
      for (const [dx, dy] of directions) {
        const nx = currentX + dx;
        const ny = currentY + dy;

        if (nx < 0 || nx >= this.width || ny < 0 || ny >= this.height) {
          hasEdge = true;
        } else {
          const neighborCell = this.board[ny][nx];
          if (neighborCell === 'B') {
            borders.add('B');
          } else if (neighborCell === 'W') {
            borders.add('W');
          } else if (neighborCell === ' ' && !visited.has(`${nx},${ny}`)) {
            visited.add(`${nx},${ny}`);
            queue.push([nx, ny]);
          }
        }
      }
    }

    const territory = Array.from(visited).map(s => s.split(',').map(Number))
      .sort((a, b) => a[0] - b[0] || a[1] - b[1]);

    if (hasEdge) {
      return { owner: 'NONE', territory };
    }

    const colors = Array.from(borders);
    if (colors.length === 1) {
      return {
        owner: colors[0] === 'B' ? 'BLACK' : 'WHITE',
        territory
      };
    } else {
      return { owner: 'NONE', territory };
    }
  }

  getTerritories() {
    const visited = new Set();
    const territories = {
      territoryBlack: [],
      territoryWhite: [],
      territoryNone: []
    };

    for (let y = 0; y < this.height; y++) {
      for (let x = 0; x < this.width; x++) {
        const key = `${x},${y}`;
        if (this.board[y][x] === ' ' && !visited.has(key)) {
          const { owner, territory } = this.getTerritory(x, y);
          territory.forEach(([tx, ty]) => {
            visited.add(`${tx},${ty}`);
          });
          if (owner === 'BLACK') {
            territories.territoryBlack.push(...territory);
          } else if (owner === 'WHITE') {
            territories.territoryWhite.push(...territory);
          } else {
            territories.territoryNone.push(...territory);
          }
        }
      }
    }

    territories.territoryBlack.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
    territories.territoryWhite.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
    territories.territoryNone.sort((a, b) => a[0] - b[0] || a[1] - b[1]);

    return territories;
  }
}
